chains/evm/voter/voter.go

```diff
@@ -28,42 +44,167 @@ type MessageHandler interface {
 	return &EVMVoter{
-		mh:             mh,
-		client:         client,
-		fabric:         fabric,
-		gasPriceClient: gasPriceClient,
+		mh:                   mh,
+		client:               client,
+		fabric:               fabric,
+		gasPriceClient:       gasPriceClient,
+		pendingProposalVotes: make(map[common.Hash]uint8),
 	}
 }
 
+// VoteProposal checks if relayer already voted and is threshold
+// satisfied and casts a vote if it isn't.
 func (v *EVMVoter) VoteProposal(m *relayer.Message) error {
 	prop, err := v.mh.HandleMessage(m)
 	if err != nil {
 		return err
 	}
-	ps, err := calls.ProposalStatus(v.client, prop)
+
+	votedByTheRelayer, err := calls.IsProposalVotedBy(v.client, v.client.RelayerAddress(), prop)
 	if err != nil {
-		return fmt.Errorf("error getting proposal: %+v status %w", prop, err)
+		return err
 	}
-	votedByTheRelayer, err := calls.IsProposalVotedBy(v.client, v.client.RelayerAddress(), prop)
+	if votedByTheRelayer {
+		return nil
+	}
+
+	shouldVote, err := v.shouldVoteForProposal(prop, 0)
 	if err != nil {
+		log.Error().Err(err)
 		return err
 	}
-	// if this relayer had not voted for proposal and proposal is in Active or Inactive status
-	// we need to vote for it
-	if !votedByTheRelayer && (ps == relayer.ProposalStatusActive || ps == relayer.ProposalStatusInactive) {
-		hash, err := calls.VoteProposal(v.client, v.fabric, v.gasPriceClient, prop)
-		log.Debug().Str("hash", hash.String()).Uint64("nonce", prop.DepositNonce).Msgf("Voted")
+
+	if !shouldVote {
+		log.Debug().Msgf("Proposal %+v already satisfies threshold", prop)
+		return nil
+	}
+
+	hash, err := calls.VoteProposal(v.client, v.fabric, v.gasPriceClient, prop)
+	if err != nil {
+		return fmt.Errorf("voting failed. Err: %w", err)
+	}
+
+	log.Debug().Str("hash", hash.String()).Uint64("nonce", prop.DepositNonce).Msgf("Voted")
+	return nil
+}
+
+// shouldVoteForProposal checks if proposal already has threshold with pending
+// proposal votes from other relayers.
+// Only works in conjuction with NewVoterWithSubscription.
+func (v *EVMVoter) shouldVoteForProposal(prop *proposal.Proposal, tries int) (bool, error) {
+	propID := prop.GetID()
+	defer delete(v.pendingProposalVotes, propID)
+
+	Sleep(time.Duration(rand.Intn(shouldVoteCheckPeriod)) * time.Second)
+
+	ps, err := calls.ProposalStatus(v.client, prop)
+	if err != nil {
+		return false, err
+	}
+
+	if ps.Status == relayer.ProposalStatusExecuted || ps.Status == relayer.ProposalStatusCanceled {
+		return false, nil
+	}
+
+	threshold, err := calls.GetThreshold(v.client, &prop.BridgeAddress)
+	if err != nil {
+		return false, err
+	}
+
+	if ps.YesVotesTotal+v.pendingProposalVotes[propID] >= threshold && tries < maxShouldVoteChecks {
+		// Wait until proposal status is finalized to prevent missing votes
+		// in case of dropped txs
+		tries++
+		return v.shouldVoteForProposal(prop, tries)
+	}
+
+	return true, nil
+}
+
+// trackProposalPendingVotes tracks pending voteProposal txs from
+// other relayers.
```

Comment:
Please, while we are here, could you please improve this description by adding more details about pending map votes